Redirect 301 for unresolved url matching with suffix

dynasource opened the pull request:
fixes #7670

dynasource committed:
create test

dynasource committed:
added fix

dynasource committed:
suggest by @samdark

samdark commented on framework/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ Yii Framework 2 Change Log
 - Bug #6363, #8301, #8582, #9566: Fixed data methods and PJAX issues when used together (derekisbusy)
 - Bug #6876: Fixed RBAC migration MSSQL cascade problem (thejahweh)
 - Bug #7627: Fixed `yii\widgets\ActiveField` to handle inputs validation with changed ID properly (dynasource, cebe)
+- Bug #7670: Fixed `yii\web\Request`: forces redirect when unresolved url matches an url with suffix (dynasource)
```
It should work the other way around as well.

dynasource replied:
there are 2 problems with this other way around:

1) The current UrlManager + UrlRules (2.0.6) accept infinit '/' suffixes in the url (try it yourself with a route like /index.php/test/route//////). This means that a suffix like '/' is always accepted in the URL, despite a suffix not being set (perhabs this is another issue for the tracker)

2) Taken this into account this means that, if you still want to check the other way around, you have to check on a '/' suffix for every request that is correctly parsed (because an given url with '////' are always parsed correctly). This goes at cost of performance.

samdark replied:
1. rtrim `/`.
2. URLs with `////` parsed correctly? What do you mean?

dynasource replied:
Dived into it. 
I started with a fresh new Advanced project with the following setting:

        'urlManager'=>[
            'class'=>'yii\web\UrlManager',
            'enablePrettyUrl' => true,            
        ]

Both following urls render correctly:
- http://localhost/dmn/yii-application/frontend/web/index.php/site/about
- http://localhost/dmn/yii-application/frontend/web/index.php/site/about/////

Seems like unwanted behavior.

SilverFire commented on framework/web/Request.php:
```diff
@@ -169,26 +169,47 @@ class Request extends \yii\base\Request
         if ($result !== false) {
-            list ($route, $params) = $result;
-            if ($this->_queryParams === null) {
+            list ($route, $params) = $result;            
```
Redundant spaces at the end of the line

SilverFire commented on framework/web/Request.php:
```diff
@@ -169,26 +169,47 @@ class Request extends \yii\base\Request
     /**
      * Resolves the current request into a route and the associated parameters.
      * @return array the first element is the route, and the second is the associated parameters.
+     * @redirect when an unresolved url matches an url with suffix 
```
`@redirect` is not a valid PHPDoc tag. Move this line above the `@return`, please.

SilverFire commented on framework/web/Request.php:
```diff
@@ -169,26 +169,47 @@ class Request extends \yii\base\Request
     {
-        $result = Yii::$app->getUrlManager()->parseRequest($this);
+        $currentQueryParams = $this->_queryParams;
+        if($url = $this->resolveRequest($this,$currentQueryParams)){
```
Here and bellow the code style must be fixed according to the PSR

`if(` -> `if (`
`){` -> `) {`

klimov-paul commented on framework/web/Request.php:
```diff
@@ -169,26 +169,47 @@ class Request extends \yii\base\Request
+        if($url = $this->resolveRequest($this,$currentQueryParams)){
+            return $url;
+        }else if(isset(Yii::$app->getUrlManager()->suffix)){
+            $request = clone $this;
```
Odd.
It does not feel right to clone `Request` instance just to check if there is a redundant/missing suffix.

dynasource replied:
well, it has been done to keep the current Request clean. When the 2nd resolve does not succeed, an error will still be thrown with the original Request (url)

klimov-paul commented on framework/web/Request.php:
```diff
@@ -169,26 +169,47 @@ class Request extends \yii\base\Request
+     * @return array the first element is the route, and the second is the associated parameters.
+     * @return false if the request cannot be resolved.
+     */
+    private function resolveRequest($request,$currentQueryParams)
```
Why passing `Request` instance as parameter to non static `Request` method?

dynasource replied:
this method is to:
- at first process the current Request 
- if not resolved, reprocess a modified Request with an altered URL (with suffix)

klimov-paul replied:
Why you can not invoke it twice like following:
```php
$this->resolveRequest();
$request->resolveRequest();
```
???

dynasource replied:
originally to create a processing method, but now I would prefer your suggestion

klimov-paul commented on framework/web/Request.php:
```diff
@@ -169,26 +169,47 @@ class Request extends \yii\base\Request
+        }else if(isset(Yii::$app->getUrlManager()->suffix)){
+            $request = clone $this;
+            $request->setPathInfo($request->getPathInfo() . Yii::$app->getUrlManager()->suffix);                
+            if($url = $this->resolveRequest($request,$currentQueryParams)){
```
This will produce significant performance impact, since URL rules will be parsed twice per each request processing.
Thus `preg_match()` calls will be doubled.

dynasource replied:
thats not correct. It will only be parsed twice if the request is NOT resolved in the first processing.

klimov-paul commented:
Solution does not seem right to me. There must be a simplier way to achieve the goal.

dynasource commented:
with the current state of the UrlManager + UrlRules + Request, this is a simple way of achieving this goal.

When you desire a cleaner and even more simple way, than more refactoring and cleanups have to be done. I would say, this is something for later releases.

dynasource committed:
Fixes based on comments:
- cleanup to PSR
- cleanup spaces
- create PHP doc return tag Response
- move Request out of resolveRequest() to $this format

SilverFire commented:
Nothing is so permanent as a temporary solution.

dynasource commented:
nothing is so permanent as a perminant 301 redirect ;)

samdark commented:
@klimov-paul do you have an idea for better approach to the problem?

df2 commented:
Just to throw it out there, we have things setup like below. This way, the default is no trailing slash but it is optional.

```php
<?php
namespace yii\helpers;

use Yii;
use yii\helpers\BaseUrl;

class Url extends BaseUrl {

    public static function to($url = '', $scheme = false)
    {
        if (is_array($url)) {

            if (!isset($url['tbs']) && $url[0] !== 'home/index') {

                $url['tbs'] = '';
            }

            return static::toRoute($url, $scheme);
        }
```

and at the end of urls.php:

```php
$newUrls = [];

foreach ($urls as $url => $route) {

    if ($url !== '/') {

        $url .= '<tbs:\/?>';
    }

    $newUrls[$url] = $route;
}

return $newUrls;
```

dynasource commented:
@df2, this patch is about automatic redirection

klimov-paul commented:
> @klimov-paul do you have an idea for better approach to the problem?

This is hard to tell for sure. But at first glance just analyzing of `$_SERVER['REQUEST_URI']` and `$_SERVER['QUERY_STRING']` should be enough to find out if suffix is present or not.

samdark commented:
Correct but we need to get the suffix we're checking for from somewhere. It could be defined either in UrlManager or in UrlRule.

klimov-paul commented:
We may anvance `UrlManaer::parseRequest()` so it can return URL suffix as a third element of the result array.
This line:
https://github.com/yiisoft/yii2/blob/master/framework/web/UrlManager.php#L268

can be following:
```php
return [(string) $route, [], $urlSuffix];
```

This will not cause a BC break.

klimov-paul commented:
Inside `Request::resolve()`, we parse 3 elements instead of 2 and so something if `$ulrSuffix` is present:

```php
public function resolve()
    {
        $result = Yii::$app->getUrlManager()->parseRequest($this);
        if ($result !== false) {
            list ($route, $params, $urlSuffix) = $result;
            
            if (isset($urlSuffix)) {
                 // check and redirect if necessary
            }

            //...
    }
```

dynasource commented:
well, it looks like your excluding the (rare) usecase of https://github.com/yiisoft/yii2/pull/9207#discussion_r36042232

Based on your example:
- you are matching by route & params first
- to get a 100% match (whether its with or without suffix)
- then you do decide to redirect

If you do this, than someone cannot have both Urls as UrlRules:
- http://exampe.com/contact/ -> pageX
- http://exampe.com/contact -> pageY

Because pageX is matched first, pageY is ignored

klimov-paul commented:
> it looks like your excluding the (rare) usecase

Why? Url suffix is used for URL rule parsing already. Does not it?
Parsing `http://exampe.com/contact/` will return result:
```php
[
    'site/contact',
    [],
    '/'
]
```
And everything will be fine.

Parsing `http://exampe.com/contact` will return
```php
[
    'site/contact',
    [],
    ''
]
```
Also should be fine.

Then we compare returned 3rd parameter with the request URI, checking if it is present or not.

dynasource commented:
the issue is about hard & softmatching and the correct order which means, first hard matching, then softmatching.

If you are doing a redirect, than it means, the URL is not right by hard matching, but with softmatching on a suffix, it should be executed.

If you want are using the current implementation of UrlManager::parseRequest, you are looping on the UrlRules. You are returning [(string) $route, [], $urlSuffix] if a match of the UrlRule has been found.

With your example, this means a softmatching on suffix is done in an early stage in the UrlRule itself, otherwise it could never return the array with the correct suffix.

This also mean that a hard match on a following UrlRule is ignored.

klimov-paul commented:
> If you are doing a redirect, than it means, the URL is not right by hard matching, but with softmatching on a suffix, it should be executed.

Are you sure about that?
UrlRule matchign includes the suffix:
https://github.com/yiisoft/yii2/blob/master/framework/web/UrlRule.php#L238

If suffix is missmatch the Rule will return `false` instead of array and thus will be skipped.

The goal is enforce some trainling suffix ("/") in case NO suffix is explicitly macthed by `Rule` and it is present in UrlManager. Thus is should be fine.

The only problem I can see if you want enforce some suffix like "/", but still keep a route without any suffix. For this use case some magic value should be reserved like `false`, indicating `Rule` should not have a suffix.

klimov-paul commented:
I have no garantee my idea will work. But at least it worth trying.

dynasource commented:
true. Your solution is cleaner & faster. If you get it more robust, fixing the problem you mentioned, I would prefer yours.

klimov-paul commented:
See #10745
